### ISSUE TITLE
Imported types don't cause typecheck error when they should.

### ISSUE THREAD

**Issue by arunabha** [facebook/flow#1528]
I'm having trouble getting imported types to work. 

In file1.js 

``` javascript
export type StringAlias = string;
```

In file2.js 

``` javascript
import type {StringAlias} from './file1.js'
var x: StringAlias = 12; 
```

Surprisingly, running flow check shows no errors !

Moving the expression in file2.js to file1.js produces the expected error from flow.

Has anyone else run into this ? Is there a special trick to getting imported types to typecheck ?

**Comment by avikchaudhuri** [comment]
You're probably forgetting to add @flow to both files (or to call check with the --all flag). Couldn't repro otherwise.

**Comment by arunabha** [comment]
@avikchaudhuri Thanks for taking a look !

Having @flow is not the issue, but you're right that recreating a minimal setup in another directory does not have this problem. 

Any theories on what might be the issue ? I've tried a bunch of tweaks to the .flowconfig file, including having just an empty one but the issue remains in one directory but not the error. Extremely frustrating :-(

**Comment by samwgoldman** [comment]
Sorry about the frustrating experience. I'd love to help you get to the bottom of this issue, but without more information it is difficult to speculate. Since the code in the original issue wasn't sufficient to reproduce, can you share more?

**Comment by arunabha** [comment]
@samwgoldman @avikchaudhuri 

After tinkering around a bit I've been able to narrow it down. Very weirdly, it looks like the problem occurs when the module exporting the type also exports an object with a parameterized function !!

It could very well be that I'm doing something wrong. A set of files illustrating the problem are attached. The flow version is 

```bash
flow version
Flow, a static type checker for JavaScript, version 0.22.1
```
[flowProblem.zip](https://github.com/facebook/flow/files/173042/flowProblem.zip)

**Comment by samwgoldman** [comment]
Thanks, @arunabha. Would you mind putting that code up as a gist instead of a zip file? Thanks.

**Comment by arunabha** [comment]
@samwgoldman  Here you go.

https://gist.github.com/arunabha/58330076b3801f4df654

**Comment by samwgoldman** [comment]
@arunabha I haven't tried to run that yet, but can you explain again what is different between `doesntWork1` and `works1` or between `doesntWork2` and `works2`? They look identical. Is it possible there's a mistake in the gist?

**Comment by arunabha** [comment]
In doesntWork1 the foo member is 

```javascript
foo: function(cb) {},
```

in works1
```javascript
foo: function() {},
```

The files that work do not export foo with a parameter, the one that don't work export foo with a parameter. I admit it sounds weird, but I just checked again :-)

**Comment by samwgoldman** [comment]
Bizarre!

**Comment by arunabha** [comment]
So, were you able to repro ? or is it just my machine / hallucination :-)

**Comment by samwgoldman** [comment]
I haven't had a chance yet, but this is high on my list!

**Comment by leonardfactory** [comment]
I don't know exactly where the problem is, however as flow requires parameter types, flow complains about `doesntWork2.js` saying:

```
  7:   foo(cb) {},
           ^^ parameter `cb`. Missing annotation
```

Adding `cb:any` i.e, `test.js` correctly reports:

```
  4: var x: StringAlias = 12;
                          ^^ number. This type is incompatible with
  4: var x: StringAlias = 12;
            ^^^^^^^^^^^ string
```

If you try to add `any` annotation, does the code typecheck?

**Comment by arunabha** [comment]
Yep, adding a ```cb: any ``` to ```doesntWork2.js``` causes ```test.js``` to start complaining about x being assigned an incompatible value.

**Comment by samwgoldman** [comment]
You *should* have gotten a missing annotation error. Did you?

**Comment by arunabha** [comment]
Yes, I did get the missing annotation error for cb in both ```doesntWork1.js``` and ```doesntWork2.js``` before adding the ```cb: any```.

Adding ```cb: any``` causes the missing annotation error to go away and the incompatible type error in test.js to show up

**Comment by samwgoldman** [comment]
Right, so Flow requires annotations on module boundaries, which is a pretty huge performance win since we can do more type checking in parallel. (We might strategically relax this restriction in the future.)

That totally explains the behavior you saw, too. If you have missing annotations like this, the module becomes `any`-typed.

**Comment by arunabha** [comment]
Hmm ok. Thanks for looking into it !

The docs do state that annotating module boundaries is a good idea, however it might be good to call out that failing to do so will make the entire module ```any```-typed. Also adding a warning about this in the output would help a lot of users.

Thanks again !

**Comment by billschaller** [comment]
This is still an issue, because flow shows test.js as having 100% coverage. 

For imported modules that flow decides to type as `any`, it should show anything referencing those modules as not covered.

```
billschalle-mbp:flow-fiddle billschaller$ flow coverage test.js
Covered: 100.00% (3 of 3 expressions)
```

**Comment by TrySound** [comment]
Closing in favour of https://github.com/facebook/flow/issues/3522
